### ISSUE TITLE
HookExtension to support high level tests

### ISSUE THREAD

**Issue by TSAVogt** [nunit/nunit#4744]
To support tests of higher test level with e.g. advanced test protocols or test environment preparation we need a reliable way to get called synchronously before and after any test relevant method (all kind of setups and teardowns, test actions and test methods).
Therefore, we would like to introduce this HookExtension interface like this:

``` csharp
public interface IHookExtension
{
    // TODO: clarfy if parameter methodName is sufficient
    // or if the (additional?) TestExecutionContext parameter is better 
    void BeforeOneTimeSetUp(string methodName);
    void AfterOneTimeSetUp(string methodName);

    void BeforeSetUp(string methodName);
    void AfterSetUp(string methodName);

    void BeforeTest(string methodName);
    void AfterTest(string methodName);

    void BeforeTearDown(string methodName);
    void AfterTearDown(string methodName);

    void BeforeOneTimeTearDown(string methodName);
    void AfterOneTimeTearDown(string methodName);
}
```
A custom extension can be activated by e.g. an IApplyToContext assembly attribute:

``` csharp
public class TestEnvironment : NUnitAttribute, IApplyToContext
{
    public void ApplyToContext(TestExecutionContext context)
    {
        context.HookExtension = new TestEnvironmentHookExtension();
    }
}
```
For more implementation details see the draft pull request: https://github.com/nunit/nunit/pull/4743

This solution would have the advantage that relevant changes are only necessary on NUnit framework, are minimal invasive and enable users to handle or orchestrate all different types of extensions for higher test levels.

What do you think about this solution?

**Comment by TSAVogt** [comment]
need some corrections first

**Comment by TSAVogt** [comment]
Just learned that I should provide some usage in the nunit.issues repo. Will provide it... Stay tuned...

**Comment by TSAVogt** [comment]
Show case is ready here: https://github.com/nunit/nunit.issues/tree/main/Issue4744

Looking forward your feedback!

**Comment by RenderMichael** [comment]
Is it possible to add hooks on each assertion? I am imagining something like middleware that's executed on each `Assert`. Ideally this would not regress the performance of the non-customized case.

**Comment by TSAVogt** [comment]
@RenderMichael: when we have finished the discussion on how to implement this featur, I will check if this mechanism can be also extended for your use case.

**Comment by TSAVogt** [comment]
As a basis for how to implement this feature, I have also added a draft PR to show what the implementation looks like if we use C# events: https://github.com/nunit/nunit/pull/4769
Simple and advanced show case: https://github.com/nunit/nunit.issues/pull/8

**Comment by TSAVogt** [comment]
According to some extensive discussions with Terje we have updated our implementation of the HookExtension for the nunit framework as well as the small demo in nunit.issues: https://github.com/nunit/nunit.issues/blob/main/Issue4744/demo.testenv/Demo.cs
Pull Request: https://github.com/nunit/nunit/pull/4769

This is a very lightweight, non-breaking and framework-only based implementation in contrast to the very intrusive implementation of additional test events that also requires changes in the engine (see https://github.com/nunit/nunit/issues/4653).

@manfred-brands @stevenaw @OsirisTerje : Could you have look at it, please? What would be necessary to bring it to mainline?

**Comment by stevenaw** [comment]
Thanks for putting this all together @TSAVogt 
I'd been trying to follow along in the discussion, and will refresh myself on everything and give this a look over in the coming days. For an extension point like this, I'd welcome a second review and perspective from another team member as well.

Appreciate how thorough and well documented your examples are as well. Thanks!

**Comment by TSAVogt** [comment]
@stevenaw: thanks a lot for your support! If you have any questions or I can help you, let me know.

**Comment by TSAVogt** [comment]
My understanding from your discussion, @stevenaw, @manfred-brands and @OsirisTerje, is that you are talking about the implementation details (events or list/implementation details). Since this issue supports other issues to be solved and I expect that this extension mechanism can be part of the NUnit framework, should I go ahead with preparing the corresponding tests as they should be independent from the implementation details?

**Comment by stevenaw** [comment]
@TSAVogt I just want to reply specifically to your question above. There seems to be a lot of interest and engagement from the team in this feature 🙂
You are welcome to prepare the corresponding tests. Thanks!

**Comment by TSAVogt** [comment]
@jnm2: First of all, sorry for the late response and thanks a lot for your input!

I tried your solution by just copying the code to a Visual Studio: worked out of the box! Very cool! 😊
And I also played around with injecting errors. I noticed some problems:

- both handlers are added as async handlers
- when catching the exception at the main level, I only get an exception and not an aggregated exception => only one of the errors is reported

But these are implementation details that can be solved for sure. I just put them here to not forget them if we go for this solution.

### Synchronous Hooks first?

Bottom line: the question was how to implement the hook extensions as synchronous calls first, so that we do not have to make breaking changes when enabling async hooks later on. You answered that well: I need to follow the EventArg pattern, which I currently do not.
==> I will adjust the pull request accordingly in the next few days like the other review comments. Thanks all for the input.

**Question to all reviewers** (@jnm2 , @stevenaw, @manfred-brands , @OsirisTerje : For the use case in my company it is absolutely fine to have "only" synchronous hooks. Is it okay with you that we start with that first? Or do you currently see a need for framework internal usage that also requires async from the beginning?

**Comment by RenderMichael** [comment]
Weighing in on the sync/async decision: 

Playwright only exposes asynchronous APIs, and has made clear that they have no intention of exposing sync versions.

Selenium is also beginning to expose async versions of their APIs. So far only page navigation is affected, but soon it will be all-encompassing. I’m not part of the inner discussions, but their tone sounds like the sync APIs will eventually be deprecated and removed.

Asynchronicity is the future of our his ecosystem, and exposing async-first APIs is the right idea in my opinion.

**Comment by TSAVogt** [comment]
Our use case is like this: we are upgrading from a very old world (customized Unit 2 + own grown test framework on top to support all test levels) and need to handle/adapt the following
![NUnit Workflow 2 HookExtensions excalidraw](https://github.com/user-attachments/assets/818aef6a-39c0-4f10-9ac8-d6567e4a381f)
The reporting hook needs to know the context so that when the test environment hook writes information, errors, ... to the log the context is needed. How to do this safely async I currently do not know (our reporting engine throws exceptions if there is something wrong with the context). 
But for sure, putting both hooks into one and shift the "special" handling from nunit framework to our side would be a way out here.
Let's wait for the feedback from the others.

**Comment by manfred-brands** [comment]
My vote is for `async` handlers. It is easy to do `sync` over `async` but not the other way around. To be clear I don't want `async void`.

The example from @jnm2 executes the handlers in parallel which assumes they are independent.
That doesn't fit with your graph. We would always run all before and after hooks.

Is the purpose only for logging status/progress? Then the failure of one hook shouldn't matter to others nor to running the actual tests.

In my opinion Hooks shouldn't affect the test. If the running of them is critical then they should be part of the test code.

**Comment by stevenaw** [comment]
re: sync vs async, I agree with @manfred-brands that a Task-based API is more flexible, especially with regards to the underlying operations performed in a handler. If reporting is a common use case then it would allow for both sync or async out-of-proc calls. It also leaves options open for if we want multiple handlers should be processed sequentially or in parallel (which seems to be a point of discussion).

I'll try to do a bit of research tonight on a few things. I believe the DataDog unit test reporter currently does a reflection-based hook to an internal method on the framework in order to report on progress. I'd like to refresh myself on how that works to have another data point for consideration here

**Comment by TSAVogt** [comment]
@manfred-brands:
Regarding purpose: it is not only logging. It is regarding having trustable test reports that are not misleading during analysis.

> In my opinion Hooks shouldn't affect the test. If the running of them is critical then they should be part of the test code.

I do not agree here. At least at high level testing. Hooks play the role of aspects for thinks like collecting relevant information in error cases, setting up and ensuring / keeping the test environment in a valid state and other needs. If they have some problem e.g. with setting up the environment it will be a waste of time to execute the test. Fail fast is here the better way.

You can also consider them as a further development of `Action` Attributes with much more hook-in-options. And with `Action` attributes it is the same: if they have a problem tests and setup methods are not executes as this this small learning test shows:

```C#
   public class TestActionThrowsException_SuiteLevel_SetUpAndTestNotExecuted
   {
       [TestsUnderTest,TestActionWithErrors]
       public class TestUnderTest
       {
           [OneTimeSetUp]
           public void OneTimeSetUp() => TestLog.LogCurrentMethod();

           [SetUp]
           public void SetUp() => TestLog.LogCurrentMethod();

           [Test]
           public void SomeTest() => TestLog.LogCurrentMethod();

           [TearDown]
           public void TearDown() => TestLog.LogCurrentMethod();

           [OneTimeTearDown]
           public void OneTimeTearDown() => TestLog.LogCurrentMethod();
       }

       [Test]
       public void CheckIt()
       {
           var testResult = TestsUnderTest.Execute();

           Assert.That(testResult.Logs, Is.EqualTo(new string[]
           {
               nameof(TestUnderTest.OneTimeSetUp),
               nameof(TestActionWithErrors.BeforeTest),
               nameof(TestActionWithErrors.AfterTest),
               nameof(TestUnderTest.OneTimeTearDown)
           }));
       }
   }
```

**Comment by TSAVogt** [comment]
@stevenaw: you are right. `sync` and `async` is maybe the wrong wording here. Sequentiell or parallel processing is the better wording. And yes, we struggle for sequentiell execution of the hooks. But that can be ensured on the user side.

=> okay you all convinced me to **go for async hook extensions** (or is `HookEvents` the better wording?). I will work on the next version of the pull request.
Thanks a lot for your input!

@stevenaw: I did a quick search for DataDog unit test reporter, but without success. If you have any details here you could also get in contact with me on Slack to avoid blowing up this thread. After our initial migration to NUnit4 we also plan to replace our home-grown test protocol by a modern reporting engine and are searching for alternatives...

**Comment by jnm2** [comment]
> And yes, we struggle for sequentiell execution of the hooks. But that can be ensured on the user side.

Yes, this is what I was going to say. If execution is required to be in a given sequential order, then the user should probably express that order explicitly by passing to AddHandler a single callback which explicitly calls each step in order. By contrast, for independent setup steps that can execute without waiting for each other, you could call AddHandler multiple times.

In the placeholder code that I shared, I want to point out that synchronous handlers are still executed in a fully synchronous way. Each one waits for the previously added handlers to complete. However, I would not rely on knowing the order in which handlers are invoked, any more than I would with C# events. As soon as order matters, I think the user should express the order explicitly within a single handler.

**Comment by stevenaw** [comment]
@TSAVogt I haven't used it myself, but it came on my radar as I saw the DataDog folks had to update integrations on their end in response to some changes we'd made here in both the 3.14 and then 4.0 releases. I had thought considering how they hook in presently could provide another use case for consideration when designing these hooks in order to maximize their use 🙂 That's said, nothing jumps out at me from looking at the source that hasn't already been discussed here.

**Comment by TSAVogt** [comment]
> > And yes, we struggle for sequentiell execution of the hooks. But that can be ensured on the user side.
> 
> Yes, this is what I was going to say. If execution is required to be in a given sequential order, then the user should probably express that order explicitly by passing to AddHandler a single callback which explicitly calls each step in order. By contrast, for independent setup steps that can execute without waiting for each other, you could call AddHandler multiple times.
> 
> In the placeholder code that I shared, I want to point out that synchronous handlers are still executed in a fully synchronous way. Each one waits for the previously added handlers to complete. (This is also true for the synchronous parts of the asynchronous handlers- that is, up until their first `await` suspension, when an unfinished Task object is returned.) However, I would not rely on knowing the order in which handlers are invoked, any more than I would with C# events. As soon as order matters, I think the user should express the order explicitly within a single handler.

Based on the feedback from others, I thought we should only provide the async API that runs in parallel. But after your answer I was not sure if we should also provide both `AddHandler` variations. I send some time for adding multiple HookExtensions to my playground project: when adding multiple assembly attributes to the `AssemblyInfo.cs` the calls to the different `ApplyToContext` methods is in reverse order compared to the order they are in the code. This means that the order in which the different hook extensions are added depends on how C# reflection works. So for me, it only makes sense to add both versions if we would also provide a possibility to control the order of execution of different `IApplyToContext` assembly attributes.

Therefore, let's keep it easy and simple and "only" add the async version. @jnm2 Do you agree?

> (All of this is discussing relative orders of handlers for the same hook. Of course, you can rely on the relative ordering of different hooks executing at different parts of the test lifecycle- for example, setup always finishes before test execution starts- the lifecycle would not continue until all handlers for a given hook have completed.)

Yes, that was an implicit precondition for me. But nevertheless, helpful to make sure that we have the same understanding. Thx 👍

**Comment by jnm2** [comment]
> So for me, it only makes sense to add both versions if we would also provide a possibility to control the order of execution of different IApplyToContext assembly attributes.

I don't agree with this. Just like with C# events, being synchronous does not imply that the relative ordering of the handlers should be reasoned about.

There may be a misunderstanding. Asynchronous does not (usually) mean using Task.Run to parallelize. Users can do that if they want, but that's a heavy-handed thing to do. You can have concurrency without multi-threading. I would strongly recommend that NUnit should execute all handlers, synchronous and asynchronous, on the same thread.

What this means is that the _awaiting_ will happen in parallel, _if there is any_. But none of the code in the asynchronous handlers will run at the same time unless the user takes things into their own hands and introduces threading.

Thus, a synchronous handler is the same as an asynchronous handler which returns `Task.Completed` at the end.

I think we should provide a synchronous overload to AddHandler, because if you don't, then people who only have a synchronous piece of work will have to use an async lambda for no reason, and get a compiler warning. And they will be able to (less efficiently) create their own synchronous AddHandler.

**Comment by jnm2** [comment]
I should add, NUnit could decide to wait for each async handler to complete before invoking the next. I'm not sure I have an opinion on that. That makes async handlers very similar to synchronous handlers. And in this case also, I would strongly make the case that the relative order of handlers should still not be relied upon. When ordering matters, relying on the execution order caused by the order of AddHandler calls introduces https://connascence.io of high strength and distant locality.

**Comment by TSAVogt** [comment]
Sorry for the delay. We are in the middle of updating the [WIP Pull Request](https://github.com/nunit/nunit/pull/4769) with the proposal of @jnm2 and setting up a test safety net for this feature. We have stumbled across an additional issue and would like to ask if you have any good ideas on how to solve it.

One of our use cases for the hook extensions is to generate test reports. This means that the `AfterTest` hook notifies our reporting engine that the test is finished. But with this call, we need to get the test outcome so that the report also contains this information. But at the moment when our AfterTest hook is called, the NUnit framework has not yet calculated the test outcome (TestResult). This is done e.g. in the `TestCommand`.

But this is only one part where the test outcome is calculated. For example, [`SimpleWorkItem.PerformWork()`](https://github.com/TSAVogt/nunit/blob/27cc9aa55e6adb472d84ba1e21e00668c30ad7ba/src/NUnitFramework/framework/Internal/Execution/SimpleWorkItem.cs#L38C1-L72C1) uses this pre-calculation and handles different scenarios where exceptions are thrown:

```C#
protected override void PerformWork()
{
    try
    {
        var testCommand = MakeTestCommand();
        Result = ContextUtils.DoIsolated(() => testCommand.Execute(Context)); // <--Here our AfterTest HookExtension is called and Result is pre-calculated
    }
    catch (Exception ex)
    {
        if (ex is ThreadAbortException)
            Thread.ResetAbort();
        Context.CurrentResult.RecordException(ex); // <-- Assertion Exceptions are recorded afterwards, so our HookExtension does not know about it when it is called.
        Result = Context.CurrentResult;
    }
    Finally
    {
        WorkItemComplete();
    }
}
```

Looking for other usages: There are about 12 other `TestCommand.Execute()` usages that handle exceptions in the above way.

We want to call an `AfterTest` hook somewhere where the test outcome has already been determined. However, in some cases we still want to be able to modify or overwrite the test outcome if necessary. 

Do you have any ideas on how to achieve this?

**Comment by TSAVogt** [comment]
For last question / problem regarding the test outcome we found a possible solution. But before opening that solution discussion I would like to close the sync/async discussion.

Therefore, @OsirisTerje, @jnm2, @manfred-brands, @stevenaw , @RenderMichael, could you have a look at the current [Pull Request](https://github.com/nunit/nunit/pull/4769) and provide feedback so that we can take the next step on the way to main branch with this enhancement.

**Comment by Sputnik24** [comment]
I just wanted to give a vote for this feature.

As far as I understand, this PR would give the option to react on start (end?) of OneTimeSetup, SetUp, TearDown and OneTimeTearDown? This is something which would help me a lot in my test logic.
